### ISSUE TITLE
Feature Request: Loading Static Resources in Jest Tests

### ISSUE THREAD

**Issue by cnaccio** [salesforce/lwc-jest#35]
Any chance lwc-jest could add support for actually loading static resources during a jest test rather than just mocking script loading functions?

I assume we'd want a library like lodash to actually be loaded during the test so that our lwc components would continue to function as expected. We may be looking at this the wrong way, but either way, any help would be much appreciated. 

In the meantime, we've settled on the solution below. 

```
jest.mock(
    'lightning/platformResourceLoader',
    () => {
        return {
            loadScript() {
                return new Promise((resolve, reject) => {
                    global.elementResizeDetectorMaker = require('../../../staticresources/elementResizeDetector');
                    resolve();
                });
            }
        };
    },
    { virtual: true }
);
```

**Comment by muenzpraeger** [comment]
Cross-referencing https://github.com/trailheadapps/lwc-recipes/issues/78

FYI @trevor-bliss

**Comment by trevor-bliss** [comment]
@cnaccio For now we've made the decision to keep the static resource imports as mocks by default. We've decided to do this for a couple reasons:

1. Similar to importing a component from another namespace, like `lightning`, it's crossing a boundary and blurs the line between a unit and higher order test. People have differing (and strong!) opinions on where exactly we should draw this boundary crossing line so I understand if you believe we should automatically import 3rd party libraries, but we've discussed internally and made the decision to default these to mocks for now.

2. We want to keep things relatively simple for what you get out of the box in `lwc-jest` and then allow more advanced users to override configuration and mocks as they please. Rather than replicating the `platformResourceLoader` logic to create a `<script>` tag with a URL that matches where your static resources are, it feels much more reliable to simply mock out the call and let users load their static resource lib themselves if they want.

FWIW I think the code sample you shared is a good approach to loading the real library. If you want a mock that is used across your entire workspace instead of per-file, you can try using using the `moduleNameMapper` config. We do that for a couple libraries in `lwc-recipes` [here](https://github.com/trailheadapps/lwc-recipes/blob/master/jest.config.js#L6).

**Comment by cnaccio** [comment]
Hey @trevor-bliss, after spending more time thinking about this and working on our tests, I'm inclined to agree with you that static resources are just like any other out scope resource for a unit test.

Plus, while at first the lack of salesforce proprietary magic bothered us, we're starting to really appreciate how vanilla the lwc dev experience is. Once we realized that lwc-jest was nothing more than  jest + some specific configuration, we were able to leverage the standard Jest documentation for troubleshooting. This is priceless!

Thanks for the response, and confirmation that our approach for static resources is solid; it's working nicely.

I'll close this issue now. Thanks again!

**Comment by AdamSquiresXAC** [comment]
Going to bump this - I feel like there is an important oversight in regards to static resources in managed packages. At present there is no method of accessing these static resources from a jest test concept, as they cannot be manually loaded through a mock without org access.

This is a deal-breaking limitation for a lot of our use-cases, as several of our packages have dependencies on JS libraries within managed packages.

Perhaps this limitation could revive the consideration of a managed static resource tool?

**Comment by trevor-bliss** [comment]
@AdamSquiresXAC Is this a managed package you also own, or one that you are only a consumer of? 

Would it be possible to see a specific example of a static resource you are consuming and how it's being used in the component? I want to make sure I fully understand the use case and how it's different from what we already discussed earlier in this thread. Feel free to simplify the example or rename things if you need to. Thank you.

**Comment by AdamSquiresXAC** [comment]
Hi Trevor, thanks for the response. Yes it's a managed package of ours. There's several re-useable JS libraries that we keep inside managed packages.

The example would be as simple as loading a JS library from a static resource within a managed package, and calling a method from the JS library. I'll try make an example today if time permits.

Thanks!

**Comment by trevor-bliss** [comment]
Hi @AdamSquiresXAC, just wanted to follow up here. To be sure I understand your use case, you own the code inside the managed package, but the source is not available unless you are inside the live org? In other words, the source is not available directly from your Salesforce DX workspace and thus you cannot import it like the example in the original post here.

If that is the case, it sounds like you need to mock out this resource since it requires org access which is out of scope for these unit tests. Or since you own the managed package, maybe there's a way to keep an in sync copy of the package contents inside your workspace that the test can import.

Let me know your thoughts and if you're unsatisfied with those answers please provide a sample project or extensive example code we can work through.

**Comment by AdamSquiresXAC** [comment]
Hi Trevor,

Thanks for getting back to me and apologies for not having the time to setup an example. Yes exactly, you've hit the nail on the head! I can see the argument to be made to keep this functionality out of scope. At the same time, it's hard to argue with the logic of:

_Re-useable code should be re-used where possible. The best model of doing this across multiple products is to store functionality in a generic package._

If the method is called only once then sure, a mock makes sense. However, when these methods are called multiple times with different arguments and expected return values, it becomes messy to do this without using the actual functionality,

I had the same thought re- keeping a local in-sync version of the external resource. The issue is that we have multiple developers working on the same projects. They could be checked out under various directory names / structures depending on what each developer is currently working on, so the unit tests would need changing depending on each developers current project directory structure.

To me this seems like a game-breaking limitation on scaleability. I think we should still have to mock these external resources, but a library to connect to an org for manually mocking external resources would solve these issues. Alternatively, a tool to temporarily download external resources from installed packages to a local directory could be an easier solution?

Perhaps a workaround would be to store the local managed package path in the project.json, and reference this url from the unit tests (if possible?). At least each developer would only need to update this URL once per directory structure, although this could become messy if developers do not keep local versions up to date (which they may not be working on for months at a time). It would also be better if project.json files could be kept insync.

Many thanks for the response and thoughts,
Adam

**Comment by muenzpraeger** [comment]
Couple of thoughts:

IMHO that's something that can be solved by an organizational level, means: defining how local directory structure should be set up by the developers. That's at least what I directed my engineers to in previous roles.

For example:
|
|-project1
|-project2
|-project3
|-shared1
|-shared2

For unit tests you can then setup a relative directory import from `projectX` to `sharedY`, similar to what we've done with the recipes [here](https://github.com/trailheadapps/lwc-recipes/blob/004aac5c49420d864ce8d3d14a4ecd74b56b897c/force-app/main/default/lwc/libsMomentjs/__tests__/libsMomentjs.test.js#L30).

Latest when you look in to incorporating unit tests into CI you'll need a similar setup, where for example a script pulls down the source of your shared code into the code that gets tested.

**Comment by trevor-bliss** [comment]
@AdamSquiresXAC Can you use an existing `sfdx` command like `sfdx force:source:pull` to pull down the package from the org and then import it in the unit test?

**Comment by AdamSquiresXAC** [comment]
@trevor-bliss no, that command will not pull down static resources from a managed package.

**Comment by trevor-bliss** [comment]
Is the source for your managed package hosted anywhere like github or npm? You could have a dependency in your project on the package directly and use it in your tests that way rather than pulling it down from the org.

To be honest I still don't think I 100% understand your use case. I understand there are nuances to managed packages but if you own the source code in the managed package it seems like there should be a way for you to have that same source available in your project during unit tests.

**Comment by AdamSquiresXAC** [comment]
Hi Trevor,

Sorry let me try and better explain. You are right in the sense that we could check out the managed package code elsewhere on our local machines, and then manually direct to these resources in jest.

The issue is that although this is a trivial thing to do for a solo developer, it is not scaleable when working in a team. Different people will have different directory structures and therefore filepaths, which would make syncing tests and config files a mess.

Does this make more sense?

**Comment by trevor-bliss** [comment]
Ok that makes sense. So going back to the question of if the managed package code is hosted anywhere like github or npm. If it's in the npm registry (or a private one only available to internal devs) then you can add a dependency on it in the package.json and it will be installed into the node_modules directory when devs run `npm install`, similar to how `lwc-jest` is loaded.

Or if you don't want to host it in the npm registry you can reference it by a github or direct URL.

**Comment by AdamSquiresXAC** [comment]
Thanks for this. We don't have anything currently hosted in npm, and if I understand correctly this would download the entire workspace of our generic package, when we are only wanting access to our static resources folder (within that directory). 

But are you saying that I can also directly access a web-hosted file within jest context? If so this makes the most sense to us! Do you have any examples of this? I've had a quick look and there's a few muddled discussion threads but it appears as though they are referring to the inclusion of local files.
